src/org/pentaho/reporting/platform/plugin/cache/FileSystemCacheBackend.java

```diff
@@ -51,50 +51,82 @@ public void setCachePath( final String cachePath ) {
 
   @Override
   public boolean write( final List<String> key, final Serializable value ) {
-    final File file = new File( cachePath + StringUtils.join( cleanKey( key ), File.separator ) );
-
-    final ObjectOutputStream oos;
-    final FileOutputStream fout;
+    final String filePath = cachePath + StringUtils.join( cleanKey( key ), File.separator );
+    final File file = new File( filePath );
+    lock( filePath );
     try {
+      //create file structure
       file.getParentFile().mkdirs();
       if ( !file.exists() ) {
         file.createNewFile();
       }
-
-      fout = new FileOutputStream( file );
-      oos = new ObjectOutputStream( fout );
-      oos.writeObject( value );
-      oos.close();
-      fout.close();
+      //closable resources
+      try (
+        final FileOutputStream fout = new FileOutputStream( file );
+        final ObjectOutputStream oos = new ObjectOutputStream( fout ) ) {
+        oos.writeObject( value );
+      }
     } catch ( final IOException e ) {
       logger.error( "Can't write cache: ", e );
       return false;
+    } finally {
+      unlock( filePath );
     }
-
     return true;
+
+  }
+
+  private void lock( final String filePath ) {
+    final File lock = new File( filePath + ".lock" );
+    while ( lock.exists() ) {
```

Comment:
And your "keep a file open for writing" allow two processes to write the same file? I dont think so.

if you follow my directions and use the user-id as one segment, your lock will only lock between the user. And if you name the lock like "${key}.lock" then your lock will be covering only the particular entry you are writing. This is as fine grained as it gets.

And remember: You need the user as ID as thanks to row-level security, the same hash can represent a report that produces different data depending on who executes it.